drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -0,0 +1,319 @@
+	u32_t idx = intid / 32;
+	u32_t val;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
+
+	val = sys_read32(ISENABLER(base, idx));
+
+	return (val & mask) != 0;
+}
+
+unsigned int arm_gic_get_active(void)
+{
+	int intid;
+
+	/* Pending -> Active or Active -> Active and Pending */
+	intid = read_sysreg(ICC_IAR1_EL1);
+
+	return intid;
+}
+
+void arm_gic_eoi(unsigned int intid)
+{
+	/* Active pending -> Active Or Active -> Inactive */
+	write_sysreg(intid, ICC_EOIR1_EL1);
+}
+
+/*
+ * Wake up GIC redistributor.
+ * clear ProcessorSleep and wait till ChildAsleep is cleared.
+ * ProcessSleep to be cleared only when ChildAsleep is set
+ * Check if redistributor is not powered already.
+ */
+static void gicv3_rdist_enable(mem_addr_t rdist)
+{
+	if (!(sys_read32(rdist + GICR_WAKER) & BIT(GIC_WAKER_CA)))
+		return;
+
+	sys_clear_bit(rdist + GICR_WAKER, GIC_WAKER_PS);
+	while (sys_read32(rdist + GICR_WAKER) & BIT(GIC_WAKER_CA))
+		;
+}
+
+/* Initialize the cpu interface */
+static void gicv3_cpuif_init(void)
+{
+	u32_t icc_sre;
+	u32_t intid;
+
+	mem_addr_t base = (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET);
+
+	/* Disable all sgi ppi */
+	sys_write32(~0, ICENABLER((GIC_RDIST_BASE + GICR_SGIBASE_OFFSET), 0));
+	/* Any sgi/ppi intid ie. 0-31 will select GICR_CTRL */
+	gic_wait_rwp(0);
+
+	/* Clear pending */
+	sys_write32(0xffffffff, ICPENDR(base, 0));
+
+	/* Configure all SGIs/PPIs as G1S.
+	 * TODO: G1S or G1NS dependending on Zephyr is run in EL1S
+	 * or EL1NS respectively.
+	 * All interrupts will be delivered as irq
+	 */
+	sys_write32(0, IGROUPR(base, 0));
+	sys_write32(0xffffffff, IGROUPMODR(base, 0));
+
+	/*
+	 * Configure default priorities for SGI 0:15 and PPI 0:15.
+	 */
+	for (intid = 0; intid < GIC_SPI_BASE; intid += 4) {
+		sys_write32(0xa0a0a0a0, IPRIORITYR(base, intid));
+	}
+
+	/* Configure PPIs as level triggered */
+	sys_write32(0, ICFGR(base, 1));
+
+	/*
+	 * Check if system interface can be enabled.
+	 * 'icc_sre_el3' needs to be configured at 'EL3'
+	 * to allow access to 'icc_sre_el1' at 'EL1'
+	 * eg: z_arch_el3_plat_init can be used by platform.
+	 */
+	icc_sre = read_sysreg(ICC_SRE_EL1);
+
+	if (!(icc_sre & BIT(SRE))) {
+		write_sysreg((icc_sre | BIT(SRE) | BIT(DIB) | BIT(DIF)),
+			     ICC_SRE_EL1);
+		icc_sre = read_sysreg(ICC_SRE_EL1);
+		assert(icc_sre & BIT(SRE));
+	}
+
+	write_sysreg(GIC_DEFAULT_PRIO, ICC_PMR_EL1);
+
+	/* Allow group1 interrupts */
+	write_sysreg(1, ICC_IGRPEN1_EL1);
+}
+
+/*
+ * TODO: Consider Zephyr in EL1NS.
+ */
+static void gicv3_dist_init(void)
+{
+	unsigned int num_ints;
+	unsigned int intid;
+	unsigned int idx;
+	mem_addr_t base = GIC_DIST_BASE;
+
+	num_ints = sys_read32(base + GIC_TYPER);
+	num_ints &= 0x1f;
+	num_ints = (num_ints + 1) << 5;
+
+	/*
+	 * Default configuration of all SPIs
+	 */
+	for (intid = GIC_SPI_BASE; intid < num_ints; intid += 32) {
+		idx = intid / NUM_INTR_PER_REG;
```

Comment:
Good catch.  Gone bad in my cleanup. And tests pass because by chance all` idx` are getting selected. 
I replaced `32` with `NUM_INTR_PER_REG`.  The define name I used is wrong.
Other places I am using ` (intid & NUM_INTR_PER_REG)  `to extract the reminder.  Will correct in next version.